homeassistant/components/google.py

```diff
@@ -0,0 +1,306 @@
+
+    oauth = OAuth2WebServerFlow(
+        config[CONF_CLIENT_ID],
+        config[CONF_CLIENT_SECRET],
+        'https://www.googleapis.com/auth/calendar.readonly',
+        'Home-Assistant.io',
+    )
+
+    persistent_notification = loader.get_component('persistent_notification')
+    try:
+        dev_flow = oauth.step1_get_device_and_user_codes()
+    except OAuth2DeviceCodeError as err:
+        persistent_notification.create(
+            hass, 'Error: {}<br />You will need to restart hass after fixing.'
+                  ''.format(err),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    persistent_notification.create(
+        hass, 'In order to authorize Home-Assistant to view your calendars'
+              'You must visit: <a href="{}" target="_blank">{}</a> and enter'
+              'code: {}'.format(dev_flow.verification_url,
+                                dev_flow.verification_url,
+                                dev_flow.user_code),
+        title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
+    )
+
+    def step2_exchange(now):
+        """Keep trying to validate the user_code until it expires."""
+        if now >= dt.as_local(dev_flow.user_code_expiry):
+            persistent_notification.create(
+                hass, 'Authenication code expired, please restart '
+                      'Home-Assistant and try again',
+                title=NOTIFICATION_TITLE,
+                notification_id=NOTIFICATION_ID)
+            listener()
+
+        try:
+            credentials = oauth.step2_exchange(device_flow_info=dev_flow)
+        except FlowExchangeError:
+            # not ready yet, call again
+            return
+
+        storage = Storage(hass.config.path(TOKEN_FILE))
+        storage.put(credentials)
+        do_setup(hass, config)
+        listener()
+        persistent_notification.create(
+            hass, 'We are all setup now. Check {} for calendars that have '
+                  'been found'.format(YAML_DEVICES),
+            title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID)
+
+    listener = track_time_change(hass, step2_exchange,
+                                 second=range(0, 60, dev_flow.interval))
+
+    return True
+
+
+def setup(hass, config):
+    """Setup the platform."""
+    config = config.get(DOMAIN, {})
+    if isinstance(config, list) and len(config) > 0:
+        config = config[0]
+
+    token_file = hass.config.path(TOKEN_FILE)
+    if not os.path.isfile(token_file):
+        do_authentication(hass, config)
+    else:
+        do_setup(hass, config)
+
+    return True
+
+
+def setup_services(hass, config):
+    """Setup service listeners."""
+    def _found_calendar(call):
+        """Check if we know about a calendar and generate PLATFORM_DISCOVER."""
+        calendar = get_calendar_info(hass, call.data)
+        if _CALENDARS.get(calendar[CONF_CAL_ID], None) is not None:
```

Comment:
We have added `hass.data` as a dictionary for components to store data. You will no longer need to use globals. Make sure you make the key unique to your component, ie `hass.data['google_calendars']`